tests/unit/test_agent.py

```diff
@@ -0,0 +1,137 @@
+# Copyright 2024 Canonical Ltd.
+# See LICENSE file for licensing details.
+#
+# Learn more about testing at: https://juju.is/docs/sdk/testing
+
+# pylint: disable=protected-access
+"""Test for agent relations."""
+
+import secrets
+from unittest.mock import MagicMock, PropertyMock
+
+import ops.testing
+import pytest
+from charms.operator_libs_linux.v1 import systemd
+
+import service
+from charm import JenkinsAgentCharm
+
+agent_relation_data = {"url": "http://example.com", "jenkins-agent-0_secret": secrets.token_hex(4)}
+
+
+def test_agent_relation_joined(harness: ops.testing.Harness):
+    """
+    arrange: initialized jenkins-agent charm.
+    act: add relation to the jenkins-k8s charm.
+    assert: The agent set the correct information in the unit's relation databag.
+    """
+    harness.begin()
+    relation_id = harness.add_relation("agent", "jenkins-k8s", unit_data=agent_relation_data)
+    charm: JenkinsAgentCharm = harness.charm
+    assert (
+        harness.get_relation_data(relation_id, app_or_unit="jenkins-agent/0")
+        == charm.state.agent_meta.as_dict()
+    )
+
+
+def test_agent_relation_changed_service_restart(
+    harness: ops.testing.Harness, monkeypatch: pytest.MonkeyPatch
+):
+    """
+    arrange: initialized jenkins-agent charm related to jenkins-k8s charm with relation data.
+    act: Trigger _on_agent_relation_changed hook.
+    assert: The charm should be in active state.
+    """
+    _ = harness.add_relation("agent", "jenkins-k8s", unit_data=agent_relation_data)
+    harness.begin()
+    charm: JenkinsAgentCharm = harness.charm
+    assert charm.state.agent_relation_credentials
+    assert (
+        charm.state.agent_relation_credentials.secret
+        == agent_relation_data["jenkins-agent-0_secret"]
+    )
+    assert charm.state.agent_relation_credentials.address == agent_relation_data["url"]
+
+    monkeypatch.setattr(charm.jenkins_agent_service, "restart", MagicMock())
```

Comment:
The `MagicMock` takes a `spec` argument which ensures that the mock is interacted with in-line with the function being mocked: https://docs.python.org/3/library/unittest.mock.html#unittest.mock.Mock. Do you think that would be helpful here?

Reply:
I thought about this for a while, but I think almost everything that is mocked for tests in this charm are methods, of which we only care about its return value and eventual side effects like raising exceptions. I'd say we can use a spec when mocking a class, but it doesn't seem to provide any additional value for a mocked method in my opinion.

Reply:
I also added the spec argument for some of the mocked class.